add mod extension for editing mod.json

NikkyAI opened the pull request:
<s>this currently only works on a existing `mod.json`

properties that can be set: `id`, `version`

planned: everything else

properties currently just override what happened to be in the mod.json, this could be changed to allow appending or prepending values
I think this would be overengineered
such things can also just stay in the `build.gradle`
</s>

new approach: dumb token replacement

available methods on the mod extension:
- `replace(String key, String value)`
- `replace(Map<String, String> replacements)`

sample usage

```kotlin
mod {
   replace("VERSION", version + "-" + buildNumber)
}
```

`mod.json`
```json
{
  "id": "fabric-language-kotlin",
  "version": "VERSION"
}
```

NikkyAI committed:
add mod extension, jar edit edition

NikkyAI committed:
check if extension has properties set inside extension

NikkyAI committed:
stupid typo

jamierocks commented on src/main/java/net/fabricmc/loom/ModExtension.java:
```diff
@@ -0,0 +1,35 @@
+import org.gradle.api.Project;
+
+public class ModExtension {
+    private final Project project;
```
Just need to fix the formatting on this file (tabs).

NikkyAI replied:
intereesting.. tabs and spaces seem to be used interchangeably in the project
net.fabricmc.loom.util.MixinRefmapHelper for example is spaces ?

NikkyAI replied:
i fixed that and all other occurences of 4 spaces that should be tabs

jamierocks reviewed (changes requested):
Looking good :+1:

NikkyAI committed:
fix other tab issues

NikkyAI commented:
sadly i cannot let idea autoformat because then it would basically touch every file.. i think someone should do that at some point though.. and maybe get a official fabric codestyle together
i suggest the intellij idea defaults with star imports banned

jamierocks commented:
There is one - see [fabric-docs](https://github.com/FabricMC/fabric-docs/blob/master/guidelines/intellij/FabricFormat.xml).

asiekierka committed:
add sources generation, update guava

asiekierka committed:
fix merged JAR generation

asiekierka committed:
switch to fernflower, add line number remapping

asiekierka committed:
download minecraft libraries through maven

asiekierka committed:
update stitch

modmuss50 committed:
Clean up GenIdeaProject, fixes run args not being updated

modmuss50 committed:
Added vscode task to generate launch.json

A few plugins will make the experience a lot better:
https://marketplace.visualstudio.com/items?itemName=redhat.java
https://marketplace.visualstudio.com/items?itemName=vscjava.vscode-java-debug
https://marketplace.visualstudio.com/items?itemName=vscjava.vscode-java-pack

asiekierka committed:
enable rebuildSourceFilenames

NikkyAI commented:
there is 2 other options we have for implementing this

:one:. do dumb string replacements
`$VERSION` -> `0.1.0-12`

:two:. pass a `Map<String, Any>`
  and handle Strings, Numbers and Maps recursively
  this options would probably be unable to cleanly extend a already existing map in the json

NikkyAI committed:
do dumb token replacement

NikkyAI commented:
i implemented option :one: for now

ChloeDawn commented:
Why is this needed for the json? Gradle provides `processResources` for stuff like this

ChloeDawn commented:
```groovy
processResources {
  from(sourceSets.main.resources.srcDirs) {
    include 'mod.json'
    expand version: project.version
  }
  from(sourceSets.main.resources.srcDirs) {
    exclude 'mod.json'
  }
}
```
```json
{
  "id": "foobar",
  "name": "FooBar",
  "version": "$version",
  "side": "universal"
}
```

NikkyAI commented:
This gets handled just fine by gradle iself just as @InsomniaKitten pointed out

so i see no reason to keep this PR open
